### ISSUE TITLE
Problem when using multiple GPUs

### ISSUE THREAD

**Issue by joeyhng** [torchnet/torchnet#56]
Sorry for posting long code here, but it is a pretty weird problem and this is the simplest example for me to possibly reproduce the problem.

I'm trying to use torchnet's ParallelDatasetIterator, but got problem when using multiple GPUs. For the following code, the output should not depend on the torchnet, since it is not using the torchnet code.

The program works normally when using single GPU:
```
$ th train_debug.lua -nGPU 1
Number of parameters: 	11184650	
iteration 1: loss=2.289696	
iteration 2: loss=2.209115	
iteration 3: loss=2.136883	
iteration 4: loss=2.071370	
iteration 5: loss=2.011046
```

However, when using 2 GPUs, the loss goes to nan after the first iteration.
```
$ th train_debug.lua -nGPU 2
Number of parameters: 	11184650	
iteration 1: loss=2.952482	
iteration 2: loss=nan	
iteration 3: loss=nan	
iteration 4: loss=nan	
iteration 5: loss=nan
```

Although the code should not really depend on torchnet, when I comment out the torchnet related code, the code works normally again.

Code:
```lua
require 'nn'
require 'cunn'
require 'cudnn'
local tnt = require 'torchnet'
local optim = require 'optim'

cmd = torch.CmdLine()
cmd:option('-nGPU', 1, 'GPU ID (only using cuda)')
cmd:option('-learning_rate', 1e-5, 'lr')
opt = cmd:parse(arg)

function dpt_model(nGPU, model)
  if nGPU > 1 then
    local gpus = torch.range(1, nGPU):totable()

    model = nn.DataParallelTable(1, true, false)
      :add(model, gpus)
      :threads(function()
        local cudnn = require 'cudnn'
      end)
    model.gradInput = nil
  end
  return model:cuda()
end


------------------------------------
-- start of torchnet related code
-- Create torchnet ParallelDatasetIterator, but not using it
local index = {}
for i=1,10000 do table.insert(index, {path='12345'}) end 
get_data_iterator_func = function ()
  return tnt.ParallelDatasetIterator{
    nthread = 8,
    init    = function() 
      require 'torchnet'
      torch.setdefaulttensortype('torch.FloatTensor')
    end,
    closure = function()
      return tnt.ListDataset{
        list = index,
        load = function (data)
          return {
            input = torch.zeros(3,224,224),
            label = torch.LongTensor{1},
          }
        end
      }:batch(16, 'skip-last')
    end
  }
end
get_data_iterator = get_data_iterator_func()
data_iterator = get_data_iterator()
-- end of torchnet related code
------------------------------------


-- modify the network a little bit (to trigger the error)
local model = torch.load('models/resnet-18.t7')
model:remove(#model)
model:add(nn.Linear(512, 10):cuda())

model = dpt_model(opt.nGPU, model)
criterion = nn.CrossEntropyCriterion():cuda()

local params, grad_params = model:getParameters()
print('Number of parameters: ', params:size(1))

-- same data in each iteration
local x = opt.nGPU > 1 and cutorch.createCudaHostTensor() or torch.CudaTensor()
local y = torch.ones(16):cuda()
x = x:resize(16,3,224,224):normal(0, 1)

function feval(xx)
  if xx ~= params then params:copy(xx) end
  grad_params:zero()
  model:training()
  local outputs = model:forward(x)
  local f = criterion:forward(outputs, y)
  local df_do = criterion:backward(outputs, y)
  model:backward(x, df_do)
  return f, grad_params
end


local optim_state = {learningRate=opt.learning_rate}
for i = 1,5 do
  local __, loss = optim.adam(feval, params, optim_state)
  cutorch.synchronize()
  print(('iteration %d: loss=%f'):format(i, loss[1]))
end
```

```models/resnet-18.t7``` is downloaded from [pretrianed model of fb.resnet.torch](https://github.com/facebook/fb.resnet.torch/tree/master/pretrained). Very weirdly, the error only triggers when the model file is put in the subdirectory. I'm able to reproduce the error in two different machines. Occasionally it runs without any problem on one of the machines that I tested on, but the problem occurs most of the time.

It may also be that I'm using DataParallelTable incorrectly. Please let me know if this is the case.

Thanks in advance!

**Comment by lvdmaaten** [comment]
I may be wrong here, but it appears you're adding the network only to the second GPU, not to the first one? Can you try something like:

```
if nGPU > 1 then
   local singlemodel = model
   local model = nn.DataParallelTable(1, true, false)
   for i = 1, nGPU do
      cutorch.setDevice(i)
      model:add(singlemodel:clone():cuda(), i)
   end
   cutorch.setDevice(1)
end
```

**Comment by joeyhng** [comment]
I think the way to use multiple GPU should be correct. I'm just following fb.resnet.torch
https://github.com/facebook/fb.resnet.torch/blob/master/models/init.lua#L81

**Comment by lvdmaaten** [comment]
I tried running the code example in your post, but I can not reproduce the issue you are seeing. Output:
```
Number of parameters: 	11184650
iteration 1: loss=2.352886
iteration 2: loss=2.275253
iteration 3: loss=2.204911
iteration 4: loss=2.140517
iteration 5: loss=2.081086
```

**Comment by lvdmaaten** [comment]
Closing this for lack of a repro or response.

**Comment by foelin** [comment]
Have you solved the problem? I also get nan when training with multiple GPU.
Thanks

**Comment by joeyhng** [comment]
No, I haven't solved the problem. I'm still interested in solving it as
well.

**Comment by ksemianov** [comment]
@joeyhng:  try `model:zeroGradParameters()` instead of `grad_params:zero()`

**Comment by joeyhng** [comment]
That seems to be the reason. It's fixed for me now. Thanks very much!!

**Comment by foelin** [comment]
it also works for me! Thanks!
